go/mysql/schema.go

```diff
@@ -56,64 +56,47 @@ CREATE TABLE if not exists _vt.schemacopy (
 	column_key varchar(3) NOT NULL,
 	PRIMARY KEY (table_schema, table_name, ordinal_position))`
 
-	detectNewColumns = `
-select ISC.table_name
-from information_schema.columns as ISC
-	 left join _vt.schemacopy as c on 
-		ISC.table_name = c.table_name and 
-		ISC.table_schema=c.table_schema and 
-		ISC.ordinal_position = c.ordinal_position
-where ISC.table_schema = database() AND c.table_schema is null`
-
-	detectChangeColumns = `
-select ISC.table_name
-from information_schema.columns as ISC
-	  join _vt.schemacopy as c on 
-		ISC.table_name = c.table_name and 
-		ISC.table_schema=c.table_schema and 
-		ISC.ordinal_position = c.ordinal_position
-where ISC.table_schema = database() 
-	AND (not(c.column_name <=> ISC.column_name) 
-	OR not(ISC.character_set_name <=> c.character_set_name) 
-	OR not(ISC.collation_name <=> c.collation_name) 
-	OR not(ISC.data_type <=> c.data_type) 
-	OR not(ISC.column_key <=> c.column_key))`
-
-	detectRemoveColumns = `
-select c.table_name
-from information_schema.columns as ISC
-	  right join _vt.schemacopy as c on 
-		ISC.table_name = c.table_name and 
-		ISC.table_schema=c.table_schema and 
-		ISC.ordinal_position = c.ordinal_position
-where c.table_schema = database() AND ISC.table_schema is null`
-
 	// DetectSchemaChange query detects if there is any schema change from previous copy.
-	DetectSchemaChange = detectChangeColumns + " UNION " + detectNewColumns + " UNION " + detectRemoveColumns
+	DetectSchemaChange = `
+SELECT DISTINCT table_name
+FROM (
+	SELECT table_name, column_name, character_set_name, collation_name, data_type, column_key
+	FROM information_schema.columns
+	WHERE table_schema = database()
+
+	UNION ALL
+
+	SELECT table_name, column_name, character_set_name, collation_name, data_type, column_key
+	FROM _vt.schemacopy c
+	WHERE table_schema = database()
+) _inner
+GROUP BY table_name, column_name, character_set_name, collation_name, data_type, column_key
+HAVING COUNT(*) = 1
```

Comment:
This new query fetches all the data from the schema copy and the `information_schema.columns` table, and uses a simple counting "trick" to determine whether a table was modified or not.

When rows that exist both in `information_schema.columns` and `_vt.schemacopy` will end up twice in the result of the `UNION ALL`, while rows that have been added / removed / modified will exist only once. Through the `HAVING COUNT(*) = 1`, we can very quickly determine the tables that have been modified.

"Look ma, no joins!"